### PR TITLE
Allow using active host context in hostfxr_get_runtime_delegate

### LINKED ISSUE #84143: Unable to use hosting APIs to load managed assemblies from an ICorProfilerCallback implementation
## Problem

Anything that isn't the host doesn't seem to be able to use the hosting APIs in order to, for example, load a managed assembly and get a function pointer as a callback. The `hostfxr_handle` required by many of the hosting APIs is not accessible by anything other than the host itself. The removal of `GetCLRRuntimeHost` will cause a loss of this functionality if the hosting APIs (or some downstream layer) do not provide some way to access the handle.

## Background

The dotnet-monitor tool uses an ICorProfilerCallback implementation in order to augment the diagnostic capabilities of the runtime. We plan to use this profiler to also load a managed assembly into the process in the case where the profiler was loaded via attach. This requires some type of API to allow the profiler to load the managed assembly and get a function callback for some function within the assembly. This can currently be done by getting the `GetCLRRuntimeHost` function and invoking it to get the `ICLRRuntimeHost*` implementation.

However, #52688 expresses a desire to remove the `GetCLRRuntimeHost` function. The alternative seems to be to use the hosting APIs to perform similar functionality via [hostfxr_get_runtime_delegate](https://github.com/dotnet/runtime/blob/main/docs/design/features/native-hosting.md#getting-a-delegate-for-runtime-functionality).

The pitfall with this function and the related hosting API functions is that it assumes that the caller has access to a `hostfxr_handle` instance. However, it seems that only the host has access to this from calling one of the `hostfxr_initialize_for_*` methods. A ICorProfilerCallback implementation, or anything else loaded into the process, is not the one that initializes the runtime, so it doesn't have access to the handle instance. As noted it #83756, invoking `hostfxr_get_runtime_delegate` without a handle will return `InvalidArgFailure (0x80008081)`.

## Possible Solutions

1. Add some way of accessing "only" (or first?) `hostfxr_handle` for the dotnet application. I believe the vast majority of dotnet executables are running a single dotnet runtime; registering and providing the single (first?) loaded `hostfxr_handle` would be convenient for others in the application to use the hosting APIs.
2. (A more scoped solution) Add a new method to a `ICorProfilerInfo*` interface that allows getting the `hostfxr_handle` for the instance of the runtime that the `ICorProfilerCallback` is attached.

cc @dotnet/dotnet-monitor, @dotnet/dotnet-diag

### PR DESCRIPTION
**Pull request by jander-msft**
These changes allow for passing `nullptr` for the handle parameter of the `hostfxr_get_runtime_delegate` function. This allows in-process callers who do not have the real host context handle to invoke this function and get a runtime delegate for the primary runtime instance. Calling `hostfxr_get_runtime_delegate` without the handle will only work if there is an active host context, otherwise it will return `StatusCode::HostInvalidState`.

For testing, I decided to add a new test specifically for ensuring that calling `hostfxr_get_runtime_delegate` without the handle will work instead of updating the existing tests. Most of the other tests would not have worked with this enablement because it requires an active context handle to be set, which only largely occurs if calling `hostfxr_get_runtime_delegate` with an existing handle or calling `fx_muxer_t::run_app`; rather than forcing these to unnecessarily invoke `hostfxr_get_runtime_delegate` twice (one with a handle, the next without), I created a separate test to do just that.

Fixes #84143

### CODE REVIEW

File: src/installer/tests/HostActivation.Tests/NativeHosting/HostContext.cs [hunk 1]
```diff
@@ -237,6 +238,31 @@ public void GetDelegate_Multiple(string checkProperties)
...
+                sharedState.RuntimeConfigPath,
+                SharedTestState.ConfigPropertyName,
+                newPropertyName
+            };
+            CommandResult result = sharedState.CreateNativeHostCommand(args, sharedState.DotNetRoot)
+                .Execute();
+
+            result.Should().Pass()
+                .And.InitializeContextForConfig(sharedState.RuntimeConfigPath)
+                .And.CreateDelegateMock_COM();
```
Comment (vitek-karas) [inline src/installer/tests/HostActivation.Tests/NativeHosting/HostContext.cs:260]: This should validate the `hdt_load_in_memory_assembly` delegate, as that is the one retrieved with `nullptr` context.

File: src/native/corehost/test/nativehost/host_context_test.cpp [hunk 2]
```diff
@@ -1000,3 +1028,13 @@ bool host_context_test::app_get_function_pointer(
 
     return app_get_function_pointer_test(hostfxr, argc, argv, config_log_prefix, test_output);
 }
+
+bool host_context_test::get_runtime_delegate(
```
Comment (vitek-karas) [inline src/native/corehost/test/nativehost/host_context_test.cpp:1032]: I would prefer to use more specific naming - `get_runtime_delegate_for_active_context` or something like that. And similarly rename the implementation method and the string passed to the helper to run it.

**Review by vitek-karas — APPROVED** [review]
Looks good - I'll leave the final call to @elinor-fung .

File: src/installer/tests/HostActivation.Tests/NativeHosting/HostContext.cs (hunk 1 above)
Comment (jander-msft) [reply src/installer/tests/HostActivation.Tests/NativeHosting/HostContext.cs]: Done

File: src/native/corehost/fxr/fx_muxer.cpp [hunk 3]
```diff
@@ -924,6 +935,29 @@ int fx_muxer_t::get_runtime_delegate(host_context_t *context, coreclr_delegate_t
...
 }
 
+int fx_muxer_t::get_runtime_delegate_active(coreclr_delegate_type type, void **delegate)
+{
+    const host_context_t *context = get_active_host_context();
+    if (context == nullptr)
+    {
+        trace::error(_X("Hosting components context has not been initialized. Cannot get runtime delegate."));
+        return StatusCode::HostInvalidState;
+    }
```
[the remark this answers is not in the document]
Comment (elinor-fung) [reply src/native/corehost/fxr/fx_muxer.cpp:945]: Perhaps Vitek means that `get_runtime_delegate` could handle not calling `load_runtime` on the const active context? You could only have something like a `const host_context_t* context_local = context == nullptr ? get_active_host_context() : context` that is used for the non-`load_runtime` operations.

Another option is to make `load_runtime` a separate function on `fx_muxer` that `hostfxr_get_runtime_delegate` would be responsible for calling if needed.

I think I have a slight preference to splitting the load and get delegate, but honestly don't feel too strongly, so I'll leave it up to you (although I would prefer the rename if you keep it as is).

File: src/native/corehost/fxr/fx_muxer.h [hunk 4]
```diff
@@ -35,6 +35,9 @@ class fx_muxer_t
         host_context_t *context,
         coreclr_delegate_type delegate_type,
         void** delegate);
+    static int get_runtime_delegate_active(
```
Comment (elinor-fung) [inline src/native/corehost/fxr/fx_muxer.h:38]: nit: I would prefer `get_runtime_delegate_for_active_context`

File: src/native/corehost/fxr/fx_muxer.cpp [hunk 5]
```diff
@@ -857,6 +857,36 @@ namespace
         g_context_initializing_cv.notify_all();
         return rc;
     }
+
+    int get_runtime_delegate_validate(const host_context_t *context, coreclr_delegate_type type)
```
Comment (elinor-fung) [inline src/native/corehost/fxr/fx_muxer.cpp:861]: nit: I'd prefer `validate_runtime_delegate_type` or `validate_delegate_type`

File: src/native/corehost/test/nativehost/host_context_test.cpp [hunk 6]
```diff
@@ -688,6 +688,34 @@ namespace
...
+        hostfxr_handle handle;
+        if (!init_for_config(hostfxr, config_path, &handle, log_prefix, test_output))
+            return false;
+
+        void *delegate1;
+        bool success = get_runtime_delegate(hostfxr, handle, delegate_type1, &delegate1, log_prefix, test_output);
+
+        // Testing that using the active host context works for get_runtime_delegate
+        // by passing nullptr for handle parameter. The runtime must be loaded for this to succeed;
+        // the first call to get_runtime_delegate with a defined handle ensures that it is loaded.
```
Comment (elinor-fung) [inline src/native/corehost/test/nativehost/host_context_test.cpp:709]: Can you also add a call using the active context before the runtime is loaded and validate that it fails as expected?

**Review by elinor-fung — APPROVED** [review]
Looks good - just some minor comments. Thank you for doing this!

File: src/native/corehost/fxr/fx_muxer.h (hunk 4 above)
Comment (jander-msft) [reply src/native/corehost/fxr/fx_muxer.h]: I've recombined the variants back into `get_runtime_delegate`, so this one is gone.

File: src/native/corehost/fxr/fx_muxer.cpp (hunk 5 above)
Comment (jander-msft) [reply src/native/corehost/fxr/fx_muxer.cpp]: I've recombined the variants back into `get_runtime_delegate`, so this one is gone.

File: src/native/corehost/test/nativehost/host_context_test.cpp (hunk 6 above)
Comment (jander-msft) [reply src/native/corehost/test/nativehost/host_context_test.cpp]: Done

File: src/native/corehost/fxr/fx_muxer.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (jander-msft) [reply src/native/corehost/fxr/fx_muxer.cpp]: Ah, not sure why I didn't think of this before. Thanks for the suggestions @vitek-karas and @elinor-fung! I've now recombined the variants back into `fx_muxer_t::get_runtime_delegate` and split `load_runtime` from this method such that `hostfxr_get_runtime_delegate` calls it if not from the active host context.

[the remark this answers is not in the document]
Comment (vitek-karas) [reply src/native/corehost/fxr/fx_muxer.cpp]: This is actually all @elinor-fung - I didn't realize that one context is mutable and the other isn't :-)
But I like the outcome.

**Comment by jander-msft** [comment]
I'm going to hazard a guess that the single failure is not really due to my change. I don't see a way of reruning that one build leg in GH or AzDO (maybe I just don't have permissions). Is there some way to do it without starting an entirely new build?

**Comment by vitek-karas** [comment]
Yes - they're unrelated. If you go to Checks and then on the left open Build Analysis, it shows that all the failures are known issues.

Thanks a lot for this PR!

**Comment by elinor-fung** [comment]
Thanks, @jander-msft!

**Pull request merged**